src/NHibernate/Linq/NhLinqExpression.cs

```diff
@@ -88,13 +88,16 @@ public IASTNode Translate(ISessionFactoryImplementor sessionFactory, bool filter
 
 			ParameterDescriptors = requiredHqlParameters.AsReadOnly();
 
-			CanCachePlan = CanCachePlan &&
-				// If some constants do not have matching HQL parameters, their values from first query will
-				// be embedded in the plan and reused for subsequent queries: do not cache the plan.
-				!ParameterValuesByName
+			if (QueryMode == QueryMode.Select && CanCachePlan)
```

Comment:
I know the cached constant bug was concerning projections, which are irrelevant to DML queries. Still this is a bit strange we need to ignored non parameterized constants: we should not have them in the first place I think. Maybe this fix dodges the real issue here: why do we have non-parameterized constants for DML queries? It is actually normal, or a bug which need to be fixed, instead of dodging the trouble?

Reply:
To me skipping unrelated logic looks more effective.

As I said here https://github.com/nhibernate/nhibernate-core/pull/2229#discussion_r328567660 it looks like intentional behavior with responsible code [here](https://github.com/nhibernate/nhibernate-core/blob/a41010f0611e93fc13c56af615b86b4bb90ad01d/src/NHibernate/Linq/DmlExpressionRewriter.cs#L103-L105). But I don't have time to investigate it thoroughly. IMHO it's not worth it.